src/vision/status_quo/barbara_benchmarks_async_trait.md

```diff
@@ -0,0 +1,134 @@
+
+Barbara sees that in a real codebase, this performance would not be good enough for writing a network stack capable of sustaining gigabit-level throughput on underpowered hardware. Barbara is disappointed, but knows that support for `async fn` in traits is in the works.
+
+Barbara looks at her organization's current priorities and decides that 100's of mbps will be an acceptable level of performance for the near term. She decides to adopt `async_trait` with the expectation that the performance benefit will go away in the long term.
+
+## 🤔 Frequently Asked Questions
+
+*Here are some standard FAQ to get you started. Feel free to add more!*
+
+### **What are the morals of the story?**
+*Talk about the major takeaways-- what do you see as the biggest problems.*
+
+* Language features that don't work well together can be a major roadblock in the course of development. Developers expect all of a language's features to be at their disposal, not using one to cut them off from using another.
+* Allocation and dynamic dispatch aren't acceptable runtime performance costs for all use cases.
+
+### **What are the sources for this story?**
+*Talk about what the story is based on, ideally with links to blog posts, tweets, or other evidence.*
+
+This story is based on actual experience implementing the 3rd-generation network stack for the Fuchsia operating system.
+
+The benchmarks are [implemented here](https://fxrev.dev/528302).
+
+### **Why do you need to model backpressure?**
+
+The Linux network stack doesn't do this; instead it drops packets as hardware buffers fill up.
+
+Because our network stack lives in a separate process from the driver, paying attention to hardware queue depth directly is not an option. There is a communication channel of bounded depth between the network stack and the driver. Dropping packets when this channel fills up would result in an unacceptable level of packet loss. Instead, the network stack must "push" this backpressure up to the applications using the network. This means each layer of the system has to be aware of backpressure.
+
+### **How would you solve this in other systems languages?**
+
+In C++ we would probably model this using callbacks which are passed all the way down the stack (through each leayer of the system).
+
+### **What's nice about async when modelling backpressure?**
+
+Futures present a uniform mechanism for communicating backpressure through polling. When requests stack up but their handler futures are not being polled, this indicates backpressure. Using this model means we get backpressure "for free" by simply adding `async` and `.await` to our code, at least in theory.
```

Comment:
Why talk about futures and polling at all? `async` functions exercise backpressure by simply "blocking" until the operation is done, which requires the necessary resources being available. So it allows to model backpressure in the same way as synchronous code  - but with the benefit of being able to run more tasks on a thread.

Reply:
I think it was a matter of not wanting to add callbacks everywhere to existing code.